crates/fluvio/src/producer/memory_batch.rs

```diff
@@ -35,7 +37,7 @@ impl MemoryBatch {
 
     /// Add a record to the batch.
     /// The value of `Offset` is relative to the `MemoryBatch` instance.
-    pub fn push_record(&mut self, mut record: Record) -> Option<Offset> {
+    pub fn push_record(&mut self, mut record: Record) -> Result<Offset, (BatchSize, Record)> {
```

Comment:
not sure why this needs to be changed? compression applies to entire batch.  adding record should not fail

Reply:
Maybe we can use an enum to be more clear. But this is not for compression, is for the cases where fluvio tries to push a record that does not fit in the batch by the batch_size. In that cases, we should create a new batch.

Reply:
that's already implemented in the higher level producer level and no need to add this level

Reply:
Sure, I adopt another strategy, instead of returning the final record size for these case when the record is larger than `batch_size`, just make optional the `batch_size` of `MemoryBatch`.

wdyt?

Reply:
it is fine as.   Only issue with returning result is that need to create specific error type to distinguish record can't be added which might be more complex than just returning option (as long as there is no other error)